File: julia/fractal/LSystem.py
from math import sin, cos, pi
import matplotlib.pyplot as pl
from matplotlib import collections

class L_System(object):
    def __init__(self, rule):
        info = rule['S']
        for i in range(rule['iter']):
            ninfo = []
            for c in info:
                if c in rule:
                    ninfo.append(rule[c])
                else:
                    ninfo.append(c)
            info = "".join(ninfo)
        self.rule = rule
        self.info = info

    def get_lines(self):
        d = self.rule['direct']
        a = self.rule['angle']
        p = (0.0, 0.0)
        l = 1.0
        lines = []
        stack = []
        for c in self.info:
            if c in "Ff":
                r = d * pi / 180
                t = p[0] + l*cos(r), p[1] + l*sin(r)
                lines.append(((p[0], p[1]), (t[0], t[1])))
                p = t
            elif c == "+":
                d += a
            elif c == "-":
                d -= a
            elif c == "[":
                stack.append((p,d))
            elif c == "]":
                p, d = stack[-1]
                del stack[-1]
        return lines


rules = [
    {
        "S":"F", "F":"F+F--F+F",
        "direct":180,
        "angle":60,
        "iter":5,
        "title":"Koch curve"
    },
    {
        "S":"F--F--F", "F":"F+F--F+F",
        "direct":180,
        "angle":60,
        "iter":5,
        "title":"Koch snowflake"
    },
    {
        "X":"X+YF+", "Y":"-FX-Y", "S":"FX",
        "direct":0,
        "angle":90,
        "iter":11,
        "title":"Dragon"
    },
    {
        "f":"F-f-F", "F":"f+F+f", "S":"f",
        "direct":0,
        "angle":60,
        "iter":7,
        "title":"Sierpinski triangle"
    },
    {
        "X":"F-[[X]+X]+F[+FX]-X", "F":"FF", "S":"X",
        "direct":-45,
        "angle":25,
        "iter":6,
        "title":"Plant"
    },
    {
        "S":"X", "X":"-YF+XFX+FY-", "Y":"+XF-YFY-FX+",
        "direct":0,
        "angle":90,
        "iter":5,
        "title":"Hilbert"
    }
]

def draw(ax, rule, iter=None):
    if iter!=None:
        rule["iter"] = iter
    lines = L_System(rule).get_lines()
    linecollections = collections.LineCollection(lines)
    ax.add_collection(linecollections, autolim=True)
    ax.axis("equal")
    ax.set_axis_off()
    ax.set_xlim(ax.dataLim.xmin, ax.dataLim.xmax)
    ax.invert_yaxis()
    
fig = pl.figure(figsize=(7,4.5))
fig.patch.set_facecolor("papayawhip")

for i in range(6):
    ax = fig.add_subplot(231+i)
    draw(ax, rules[i])

fig.subplots_adjust(left=0,right=1,bottom=0,top=1,wspace=0,hspace=0)
pl.show()

File: julia/fractal/LSystem.jl
using Cairo

# LSystem 
const Rule = Dict{String, Any}
const RULES = [
Dict(
    "S"=>"F", "F"=>"F+F--F+F",
    "direct"=>0,
    "angle"=>60,
    "iter"=>5,
    "title"=>"Koch curve"
),
Dict(
    "S"=>"F--F--F", "F"=>"F+F--F+F",
    "direct"=>0,
    "angle"=>60,
    "iter"=>5,
    "title"=>"Koch snowflake"
),
Dict(
    "S"=>"FX", "X"=>"X+YF+", "Y"=>"-FX-Y",
    "direct"=>0,
    "angle"=>90,
    "iter"=>11,
    "title"=>"Dragon"
),
Dict(
    "S"=>"f", "f"=>"F-f-F", "F"=>"f+F+f", 
    "direct"=>0,
    "angle"=>60,
    "iter"=>7,
    "title"=>"Sierpinski triangle"
),
Dict(
    "S"=>"X", "X"=>"F-[[X]+X]+F[+FX]-X", "F"=>"FF", 
    "direct"=>-45,
    "angle"=>25,
    "iter"=>6,
    "title"=>"Plant"
),
Dict(
    "S"=>"X", "X"=>"-YF+XFX+FY-", "Y"=>"+XF-YFY-FX+",
    "direct"=>90,
    "angle"=>90,
    "iter"=>6,
    "title"=>"Hilbert"
),
]

const DEFAULT_RULE = RULES[1]

function genStepString(rule::Rule) :: String
    start = rule["S"]
    iter  = rule["iter"]
    angle = rule["angle"]
    
    for i in 1:iter
        rt = []
        for c in start
            c = string(c)
            push!(rt, get(rule, c, c))
        end
        start = join(rt, "")
    end
    
    start
end

function genStep(rule::Rule=DEFAULT_RULE)
    angle = rule["direct"]
    Δθ = rule["angle"]
    sp = (0, 0) # start point
    l = 10 # step length
    
    steps = []
    stack = []
    str = genStepString(rule)
    
    for c in str
        if c in "Ff"
            r = angle * 2π/360
            ep = sp .+ (l*cos(r), l*sin(r))
            push!(steps, (sp, ep))
            sp = ep
        elseif c == '+'
            angle += Δθ
        elseif c == '-'
            angle -= Δθ
        elseif c == '['
            push!(stack, (sp, angle))
        elseif c == ']'
            sp, angle = pop!(stack)
        else
            # @warn("Invalid character `$c` in steps string.")
        end
    end
    
    steps
end

# -------------------------------------------------
W = 512
H = W

dx = H/2
dy = dx


c = CairoRGBSurface(W, H);
cr = CairoContext(c);

save(cr);
set_source_rgb(cr,0.8,0.8,0.8);    # light gray
rectangle(cr,0.0,0.0,W,H); # background
fill(cr);
restore(cr);

# draw helping lines 
set_source_rgb(cr, 1, 0.2, 0.2);
set_line_width(cr, 1);

move_to(cr, 0.0+dx, 0.0+dy);
steps = genStep(RULES[4])
for st in steps
    x, y = st[2]
    line_to(cr, x+dx, y+dy);
end
stroke(cr);

# write_to_png(c,"sample_set_line_cap.png");
c
